dspy/functional/functional.py

```diff
@@ -98,12 +98,20 @@ def __init__(self, signature, instructions=None, *, max_retries=3, wrap_json=Fal
             explain_errors: If True, the model will try to explain the errors it encounters.
         """
         super().__init__()
-        self.signature = ensure_signature(signature, instructions)
+        signature = ensure_signature(signature, instructions)
         self.predictor = dspy.Predict(signature, _parse_values=False)
         self.max_retries = max_retries
         self.wrap_json = wrap_json
         self.explain_errors = explain_errors
 
+    @property
+    def signature(self) -> dspy.Signature:
+        return self.predictor.signature
+
+    @signature.setter
```

Comment:
If you want to ensure that this value is not set outside the object initialization (bootstrapping), I would remove the getter to prevent anyone from using it to set the value. Of course, the tests should still pass.

In summary, if you don't plan to use it, the setter is unnecessary.

Reply:
Unfortunately we have to keep `self.signature` being readable and settable, otherwise it could break people's existing code.